### PR TITLE
Feat/m pbrs

### PR DESCRIPTION
**Pull request by Abhishekism9450**

### CODE REVIEW

File: examples/pbr/main.tf [hunk 1]
```diff
@@ -0,0 +1,152 @@
+terraform{
+    required_providers {
+        nutanix = {
+            source  = "nutanixtemp/nutanix"
+            version = "1.99.99"
```
Comment (bhati-pradeep) [inline examples/pbr/main.tf]: Change to next release version once decided

File: examples/pbr/main.tf [hunk 2]
```diff
@@ -0,0 +1,152 @@
+terraform{
+    required_providers {
+        nutanix = {
+            source  = "nutanixtemp/nutanix"
```
Comment (bhati-pradeep) [inline examples/pbr/main.tf]: nutanix/nutanix

File: nutanix/data_source_nutanix_pbr.go [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+										Type:     schema.TypeList,
+										Computed: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"tcp": {
+													Type:     schema.TypeList,
+													Computed: true,
+													Elem: &schema.Resource{
+														Schema: map[string]*schema.Schema{
+															"destination_port_range_list": portRangeSchema(),
```
Comment (bhati-pradeep) [inline nutanix/data_source_nutanix_pbr.go]: portRangeSchema contains both Optional and Computed as true. I believe its for creating a common func for resource and datasource both. @siddharth-kulshrestha should we go ahead with this or theres should be two separate schemas for each ?

File: nutanix/data_source_nutanix_pbr.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+													Type:     schema.TypeString,
+													Computed: true,
+												},
+											},
+										},
+									},
+									"protocol_type": {
+										Type:     schema.TypeString,
+										Computed: true,
+									},
```
Comment (bhati-pradeep) [inline nutanix/data_source_nutanix_pbr.go:305]: If this field can only be set with certain limited values then we can add a validator here to throw error incase user gives wrong value. Check this https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#:~:text=%C2%BB-,ValidateDiagFunc,-Data%20structure%3A

File: nutanix/data_source_nutanix_pbr.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+func dataSourceNutanixPbrRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*Client).API
+
+	pbrUUID, ok := d.GetOk("pbr_uuid")
+	if !ok {
+		return diag.Errorf("please provide pbr reference uuid")
+	}
+
```
Comment (bhati-pradeep) [inline nutanix/data_source_nutanix_pbr.go:323]: Doesnt api support query with name as well ? Or any other API we can use incase to query using name

File: nutanix/resource_nutanix_pbr.go [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"api_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"metadata": {
```
Comment (bhati-pradeep) [inline nutanix/resource_nutanix_pbr.go:173]: Can we attach categories to PBRs. If yes, then please include them as well.

File: nutanix/resource_nutanix_pbr.go [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+		_, n := d.GetChange("protocol_parameters")
+		res.ProtocolParameters = expandProtocolParameters(n.([]interface{}))
+	}
+
+	if d.HasChange("action") {
+		actUpdated := &v3.PbrAction{}
+		action := d.Get("action")
+		actUpdated.Action = utils.StringPtr(action.(string))
+
+		if sip, sok := d.GetOk("service_ip_list"); sok {
```
Comment (bhati-pradeep) [inline nutanix/resource_nutanix_pbr.go:371]: Please check if utils have any method for this kind of conversion below

File: nutanix/resource_nutanix_pbr.go [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+		Pending:    []string{"PENDING", "RUNNING"},
+		Target:     []string{"SUCCEEDED"},
+		Refresh:    taskStateRefreshFunc(conn, taskUUID),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      PbrDelayTime,
+		MinTimeout: PbrMinTimeout,
+	}
+
+	if _, errWaitTask := stateConf.WaitForStateContext(ctx); errWaitTask != nil {
+		return diag.Errorf("error waiting for PBR (%s) to create: %s", d.Id(), errWaitTask)
```
Comment (bhati-pradeep) [inline nutanix/resource_nutanix_pbr.go]: to update

File: nutanix/resource_nutanix_pbr.go [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+
+		act.Action = utils.StringPtr(action.(string))
+
+		if sip, sok := d.GetOk("service_ip_list"); sok {
+			subips := sip.([]interface{})
+			sublist := make([]string, len(subips))
+			for a := range subips {
+				sublist[a] = subips[a].(string)
+			}
+			act.ServiceIPList = sublist
```
Comment (bhati-pradeep) [inline nutanix/resource_nutanix_pbr.go:494]: Same. Check utils for this conversion functionality

File: nutanix/data_source_nutanix_pbr.go (hunk 5 above)
Comment (bhati-pradeep) [reply nutanix/data_source_nutanix_pbr.go:323]: Can we use filter query from list api ?

File: nutanix/data_source_nutanix_pbr.go (hunk 4 above)
Comment (Abhishekism9450) [reply nutanix/data_source_nutanix_pbr.go:305]: validator is there. Check latest commit .

File: nutanix/resource_nutanix_pbr.go (hunk 6 above)
Comment (Abhishekism9450) [reply nutanix/resource_nutanix_pbr.go:173]: No categories present in PBR .

File: nutanix/resource_nutanix_pbr.go (hunk 9 above)
Comment (Abhishekism9450) [reply nutanix/resource_nutanix_pbr.go:494]: no util is present to convert list of strings to list of strings.

File: nutanix/resource_nutanix_pbr.go (hunk 6 above)
Comment (Abhishekism9450) [reply nutanix/resource_nutanix_pbr.go:173]: No categories present in PBR resource.

**Review by siddharth-kulshrestha — APPROVED** [review]

File: nutanix/data_source_nutanix_pbr.go (hunk 5 above)
Comment (Abhishekism9450) [reply nutanix/data_source_nutanix_pbr.go:323]: no , api doesn't support query with name. I have introduced filter incase if any other attribute supports.

**Pull request merged**